components/calendar/src/japanese.rs

```diff
@@ -215,6 +247,103 @@ impl Calendar for Japanese {
     }
 }
 
+impl Date<Japanese> {
+    /// Construct a new Japanese Date.
+    ///
+    /// Years are specified in the era provided, and must be in range for Japanese
+    /// eras (e.g. dates past April 30 Heisei 31 must be in Reiwa; "Jun 5 Heisei 31" and "Jan 1 Heisei 32"
+    /// will not be adjusted to being in Reiwa 1 and 2 respectively)
+    ///
+    /// However, dates may always be specified in "bce" or "ce" and they will be adjusted as necessary.
+    ///
+    /// ```rust
+    /// use icu::calendar::{types, Date, Ref};
+    /// use icu::calendar::japanese::{Japanese, JapaneseEraStyle};
+    /// use std::convert::TryFrom;
+    /// use tinystr::tinystr;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    /// let japanese_calendar = Japanese::try_new(&provider, JapaneseEraStyle::Modern).expect("Cannot load japanese data");
+    /// // for easy sharing
+    /// let japanese_calendar = Ref(&japanese_calendar);
+    ///
+    /// let era = types::Era(tinystr!(16, "heisei"));
+    ///
+    /// let date = Date::new_japanese_date(japanese_calendar, era, 14, 1, 2)
+    ///     .expect("Constructing a date should succeed");
+    ///
+    /// assert_eq!(date.year().era, era);
+    /// assert_eq!(date.year().number, 14);
+    /// assert_eq!(date.month().ordinal, 1);
+    /// assert_eq!(date.day_of_month().0, 2);
+    ///
+    /// // This function will error for eras that are out of bounds:
+    /// // (Heisei was 32 years long, Heisei 33 is in Reiwa)
+    /// let oob_date = Date::new_japanese_date(japanese_calendar, era, 33, 1, 2);
+    /// assert!(oob_date.is_err());
+    ///
+    /// // and for unknown eras
+    /// let fake_era = types::Era(tinystr!(16, "neko")); // 🐱
+    /// let fake_date = Date::new_japanese_date(japanese_calendar, fake_era, 10, 1, 2);
+    /// assert!(fake_date.is_err());
+    /// ```
+    pub fn new_japanese_date<A: AsCalendar<Calendar = Japanese>>(
```

Comment:
One thing I'm not sure about: Should the calendar be the _first_ parameter on these functions or the last?

`Date::new_from_codes()` has it as the last one, this has it as the first, I'm not sure if I like either better. We should probably be uniform.

Reply:
The policy we've generally been leaning toward is that parameters that are common across all the functions go first, and the less common ones go later. Data provider is more common than locale which is more common than options.

Reply:
Okay, by that rule `new_japanese_date()` should have the calendar at the end.

Should `new_from_codes()` and `new_from_iso()` do the same? Note that params are the same for every specialization of both those functions

Reply:
Ah, they already do the same. I think this is good then.

Reply:
I didn't mean to imply that the "policy we've generally been leaning toward" is my preference for this particular instance. :)

The `calendar` is a common argument in several of these functions. It is also something that provides "context" as opposed to the specific "options", which are the era/year/month/day.

But I don't have a very strong opinion.

Reply:
Yep, that convinces me. It's already at the end for all of them, resolving this.